### DISCUSSION TITLE
Using AI-deck to control Crazyflie

### DISCUSSION THREAD

**Discussion by Yingxiu-Chang** [bitcraze/discussions#874] [Q&A] [answered]
Hi, 
I have a similar tflite model like the classification demo from https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/ai-examples/classification-demo/. I want to use this model to send the commands to the crazyflie via UART. Does it have any examples or support documents about the communicaion between the AI-deck and Crazyflie?
Best regards
Hugh

**Comment by knmcguire** [comment]
Hi!

We would recommend you using the CPX protocol for that https://www.bitcraze.io/documentation/repository/aideck-gap8-examples/master/infrastructure/cpx/.

Make sure to read it first, as the app layer variant hasn't been implemented yet but all the framework is there that should be low effort to implement it if you have some experience with embedded programming.

**Reply by Yingxiu-Chang** [reply]
Hi, 
Thank you for your prompt responses. 
I also tried to compile and test the uart_send_char by typing: 'sudo docker run --rm -v ${PWD}:/module aideck-with-autotiler tools/build/make-example examples/other/test_functionalities/uart_send_char image'. 

It fails and shows 

```
Collecting numpy==1.22.3
Using cached numpy-1.22.3-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.8 MB)
ERROR: imageio 2.17.0 has requirement pillow>=8.3.2, but you'll have pillow 7.2.0 which is incompatible.
Installing collected packages: numpy
Attempting uninstall: numpy
Found existing installation: numpy 1.19.5
Uninstalling numpy-1.19.5:
Successfully uninstalled numpy-1.19.5
Successfully installed numpy-1.22.3
gapy --target=ai_deck --platform=board --work-dir=/module/examples/other/test_functionalities/uart_send_char/BUILD/GAP8_V2/GCC_RISCV_PULPOS run --image --binary=/module/examples/other/test_functionalities/uart_send_char/BUILD/GAP8_V2/GCC_RISCV_PULPOS/test
usage: gen_flash_image [-h]
[--target {ai_deck,gapoc_b,gapuino_v2_spi,gapoc_a_v2_spi,gapuino_v2,gap9_v2,gapuino,pulp,gapuino_spi,gapoc_a_v2,pulp_rnnext,gapoc_b_v2,gapoc_a,gapuino_v3,gapuino_v3_spi,gapoc_b_v2_spi,pulp_nn,gap9}]
[--py-target PY_TARGET]
[--platform {board,rtl,gvsoc,fpga,netlist,pgpin}] [-v]
[--work-dir WORK_DIR] [--config-ini INI_CONFIGS]
[--config-opt CONFIG_ITEMS] [--boot-loader BOOT_LOADER]
[--partition-table PARTITIONTABLE]
[--partition partition-name partition-image partition-name partition-image]
[--block-size BLOCKSIZE] [--flash-size FLASHSIZE]
[--flash-type {hyper,spi}]
[--flash-stimuli-format {s,l,m}]
[--flash-stimuli-file {s,l,m}] [--encryte ENCRYPT]
[--aes_key AES_KEY] [--aes_iv AES_IV] [-o OUTPUT]
gen_flash_image: error: argument --boot-loader: can't open '/module/examples/other/test_functionalities/uart_send_char/BUILD/GAP8_V2/GCC_RISCV_PULPOS/test': [Errno 2] No such file or directory: '/module/examples/other/test_functionalities/uart_send_char/BUILD/GAP8_V2/GCC_RISCV_PULPOS/test'
make: *** [/gap_sdk/utils/rules/pulp_rules.mk:243: image] Error 2
```

I don't have too much experiences on this platform. 
Would you please tell me how to compile the example of 'uart_send_char'?
Best regards
Hugh

**Reply by knmcguire** [reply]
This is actually a non-cpx example, so this is of a low level char sending example and perhaps not really useful.

Is the file being created in the directory of the '/examples/other/test_functionalities/uart_send_char/BUILD/GAP8_V2/GCC_RISCV_PULPOS/test' ?

**Reply by Yingxiu-Chang** [reply]
It cannot be created. 
I'm trying to create a packet on AI-deck and send the packet to crazyflie via CPX. However, most of the examples used cpxPrintToConsole. 
Do you have any C code like using a counter or anything, then send a value via CPX instead of cpxPrintToConsole?
Best regards
Hugh

**Comment by krichardsson** [comment]
I think you can look at the [`hello_world_gap8.c` example](https://github.com/bitcraze/aideck-gap8-examples/blob/master/examples/other/hello_world_gap8/hello_world_gap8.c) to get a bit more understanding of CPX. (Also read the [documentation](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/cpx/))

The [`cpxPrintToConsole()`](https://github.com/bitcraze/aideck-gap8-examples/blob/fdb3b49f1f390f4b30577b6098a6a976e16c0a39/lib/cpx/src/cpx.c#L102-L120) is just doing a bit more than you want, but you can look at it to see how it is implemented. The short version is: 
1. Set up the header (source, destination and so on)
2. Add the text as data
3. Call `cpxSendPacketBlocking()`

The preferred way to set up the header information is to use [`cpxInitRoute()`](https://github.com/bitcraze/aideck-gap8-examples/blob/fdb3b49f1f390f4b30577b6098a6a976e16c0a39/lib/cpx/src/cpx.c#L122C1-L128) (which is not used in cpxPrintToConsole()!)
To send the CPX packet to a Crazyflie app (running in the STM) the `destination` should be `CPX_T_STM32` and the `function` should be `CPX_F_APP`.

To tie things together, you can see in the calls to `cpxPrintToConsole()`, `LOG_TO_CRTP` is used as the target, but this is defined as `CPX_T_STM32` so it actually sends the packet to the STM. The difference is that `function` is set to `CPX_F_CONSOLE` which will route it to a function that prints the packet to the console. 

In the Crazyflie app (in the STM) you will receive the packet through a callback. To do this
1. Create a function `void myCpxPacketCallback(const CPXPacket_t* cpxRx) { // Do stuff...}`
2. Register the callback by calling `cpxRegisterAppMessageHandler(myCpxPacketCallback);`

When a CPX packet is sent from the GAP8 it will be routed through the ESP32 to the STM and finally be delivered to you as a call to the `myCpxPacketCallback()` funciton.

**Comment by Yingxiu-Chang** [comment]
Hi, 
Thank you so much for your constructive help. Will take a try.
best
Hugh

**Comment by krichardsson** [comment]
I'm adding some example code for communicating between the Crazyfile and GAP8. It is still PRs but will soon be merged.
Please see
https://github.com/bitcraze/aideck-gap8-examples/pull/134
https://github.com/bitcraze/crazyflie-firmware/pull/1311

**Reply by Yingxiu-Chang** [reply]
Hi @krichardsson 
Thank you so much for your help. Your codes all work well and target exactly to what I need. 
Best wishes
Hugh

**Reply by Yingxiu-Chang** [reply]
Hi @krichardsson ,
A quick question. Your code stm_gap8_cpx.c is sending uint_t data[0] = counter, but dataLength is 1. 
If I want to change data to float, like data = 1.30 + 0.01 as example, then what I need to change in your code? dataLength? or anything else?
Best regards
Hugh

**Reply by krichardsson** [reply]
You can do something like this:
```C
    float myData = 47.11;
    memcpy(&txPacket.data[0], &myData, sizeof(myData));
    txPacket.dataLength = sizeof(myData);
```

**Reply by Yingxiu-Chang** [reply]
Hi @krichardsson ,
Much appreciate for your help. I changed the stm_gap8_cpx.c on crazyflie firmware based on your instructions as followings:
```
#include <string.h>
#include <stdint.h>
#include <stdbool.h>

#include "app.h"

#include "cpx.h"
#include "cpx_internal_router.h"

#include "FreeRTOS.h"
#include "task.h"

#define DEBUG_MODULE "APP"
#include "debug.h"

// Callback that is called when a CPX packet arrives
static void cpxPacketCallback(const CPXPacket_t* cpxRx);

static CPXPacket_t txPacket;

void appMain() {
  DEBUG_PRINT("Hello! I am the stm_gap8_cpx app\n");

  // Register a callback for CPX packets.
  // Packets sent to destination=CPX_T_STM32 and function=CPX_F_APP will arrive here
  cpxRegisterAppMessageHandler(cpxPacketCallback);

  //uint8_t counter = 0;
  float counter = 47.11;
  while(1) {
    vTaskDelay(M2T(2000));

    cpxInitRoute(CPX_T_STM32, CPX_T_GAP8, CPX_F_APP, &txPacket.route);
    //txPacket.data[0] = counter;
    memcpy(&txPacket.data[0], &counter, sizeof(counter));
    txPacket.dataLength = sizeof(counter);

    cpxSendPacketBlocking(&txPacket);
    DEBUG_PRINT("Sent packet to GAP8 (%f)\n", (double)counter);
    //counter++;
  }
}

static void cpxPacketCallback(const CPXPacket_t* cpxRx) {
  DEBUG_PRINT("Got packet from GAP8 (%f)\n", (double)cpxRx->data[0]);
}
```
Then I the stm_gap8_cpx.c on aideck-examples as followings:
```
#include "pmsis.h"
#include "bsp/bsp.h"
#include "cpx.h"

static CPXPacket_t rxPacket;
static CPXPacket_t txPacket;

void start_example(void) {
  pi_bsp_init();
  cpxInit();
  cpxEnableFunction(CPX_F_APP);

  cpxPrintToConsole(LOG_TO_CRTP, "Starting counter bouncer\n");

  while (1) {
    cpxReceivePacketBlocking(CPX_F_APP, &rxPacket);
    float counterInStm = rxPacket.data[0];

    cpxPrintToConsole(LOG_TO_CRTP, "Got packet from the STM (%f)\n", (double)counterInStm);

    // Bounce the same value back to the STM
    cpxInitRoute(CPX_T_GAP8, CPX_T_STM32, CPX_F_APP, &txPacket.route);
    txPacket.data[0] = counterInStm;
    txPacket.dataLength = sizeof(counterInStm);

    cpxSendPacketBlocking(&txPacket);
  }
}

int main(void) {
  return pmsis_kickoff((void *)start_example);
}
```
After flashing and open the console on console. I got followings:
```
APP: Sent packet to GAP8 (47.110000)
CPX: GAP8: Got packet from the STM (164.000000)
APP: Got packet from GAP8 (164.0)
APP: Sent packet to GAP8 (47.110000)
CPX: GAP8: Got packet from the STM (164.000000)
APP: Got packet from GAP8 (164.0)
APP: Sent packet to GAP8 (47.110000)
CPX: GAP8: Got packet from the STM (164.000000)
APP: Got packet from GAP8 (164.0)
APP: Sent packet to GAP8 (47.110000)
CPX: GAP8: Got packet from the STM (164.000000)
APP: Got packet from GAP8 (164.0)
APP: Sent packet to GAP8 (47.110000)
CPX: GAP8: Got packet from the STM (164.000000)
APP: Got packet from GAP8 (164.0)
```
I'm really new to the crazyflie. I want to modify the 'GAP8: Got packet from the STM (164.000000)' to 'GAP8: Got packet from the STM (47.11000)' and change 'Got packet from GAP8 (164.0)' to 'Got packet from GAP8 (47.11000)'. 

Will be really appreciate if you can check my code and tell me how to modify it. 
Best regards
Hugh

**Reply by Yingxiu-Chang** [reply]
I think I got your point. Do you mean to define a function to change the uint_8 to float?
Best regards
Hugh

**Reply by krichardsson** [reply, marked as answer]
On the receiving side you need to something like this:
```C
  float theResult = 0.0f;
  memcpy(&theResult, &cpxRx->data[0], sizeof(theResult));
```

> Do you mean to define a function to change the uint_8 to float? 

It is possible that you cast the data in the buffer to a float, but it is perhaps not guaranteed that the buffer is word aligned, so a memcpy() is safer. 

Also, if you want to print a float in the STM, you need to cast it to a double.
```C
  DEBUG_PRINT("Got packet from GAP8 (%f)\n", (double)theResult);
```

I really think you should read a book about the C-language, there are a lot of pitfalls and it is hard to just winging it.

**Reply by Yingxiu-Chang** [reply]
Hi, 
Thanks a lot for your help and really appreciate your suggestions.
Best wishes
Hugh
